File: generator/src/emitters/branch_skip.cpp
#include "../state.hpp"
#include "branch_skip.hpp"
#include <array>
#include <format>
#include <random>
#include <string_view>

namespace stim {
std::optional<std::string> BranchSkipEmitter::try_emit(GeneratorState& state) {
  auto src = state.pick_initialized_gpr();
  if (!src) return std::nullopt;

  static constexpr std::array<std::string_view, 2> ops = {"CBZ", "CBNZ"};
  std::uniform_int_distribution<size_t> op_dist(0, 1);
  const auto op = ops[op_dist(state.rng())];

  const uint32_t n = state.next_label_id();
  state.hit_coverage(CovPoint::BranchDense);
  return std::format("    {} x{}, .Lskip_{}\n    NOP\n.Lskip_{}:", op, *src, n, n);
}
} // namespace stim

File: rust-generator/src/emitters/branch_skip.rs
use crate::coverage::CovPoint;
use crate::emitter::Emitter;
use crate::state::GeneratorState;
use rand::RngExt;

pub struct BranchSkipEmitter;

impl Emitter for BranchSkipEmitter {
    fn try_emit(&mut self, state: &mut GeneratorState) -> Option<String> {
        let src = state.pick_initialized_gpr()?;
        const OPS: [&str; 2] = ["CBZ", "CBNZ"];
        let op = OPS[state.rng().random_range(0..2)];
        let n = state.next_label_id();
        state.hit_coverage(CovPoint::BranchDense);
        Some(format!(
            "    {} x{}, .Lskip_{}\n    NOP\n.Lskip_{}:",
            op, src, n, n
        ))
    }
}
